Take #14026 and #14109 from upstream

masonh22 opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #.

 # Rationale for this change
<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

# What changes are included in this PR?
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->
Takes #14026 and #14109 from upstream

This PR mirrors the #17757 upstream PR.

# Are there any user-facing changes?
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

nuno-faria committed:
feat(optimizer): Enable filter pushdown on window functions (#14026)

* feat(optimizer): Enable filter pushdown on window functions

Ensures selections can be pushed past window functions similarly
to what is already done with aggregations, when possible.

* fix: Add missing dependency

* minor(optimizer): Use 'datafusion-functions-window' as a dev dependency

* docs(optimizer): Add example to filter pushdown on LogicalPlan::Window

(cherry picked from commit ad5a04f47f67090b6acb2ddd46b5f15200e85912)

berkaysynnada committed:
Avoid Aliased Window Expr Enter Unreachable Code (#14109)

(cherry picked from commit fda500ab8a5b4232a6619ad9875f99cf97b95c5c)

masonh22 committed:
Remove tests that weren't part of the commit but were added by the cherry pick

masonh22 committed:
Fix test trainings (remove space in PhysicalSortExpr)

avantgardnerio reviewed (approved):
LGTM. Before merge, please:
1. Try the `rev` for this PR in DQE `Cargo.toml` & test
2. See if you can disable the failing checks (if that doesn't work we can have Nathaniel do it)

masonh22 committed:
Use `Expr::qualified_name()` and `Column::new()` to extract partition keys.

Using `Expr::schema_name()` and `Column::from_qualified_name()` could
incorrectly parse the column name.

avantgardnerio commented:
It looks like this also contains #14109 ? Could you add that to the description please?

masonh22 committed:
Revert "Fix test trainings (remove space in PhysicalSortExpr)"

This reverts commit 3bda922d0bc9f47fc3c52f63139a45faa6917e7c.

masonh22 committed:
Add the test from #17757
